platforms/core-configuration/model-core/src/main/java/org/gradle/internal/snapshot/impl/IsolatedArrayOfPrimitive.java

```diff
@@ -0,0 +1,166 @@
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.snapshot.impl;
+
+import org.gradle.api.NonNullApi;
+import org.gradle.internal.hash.Hasher;
+import org.gradle.internal.isolation.Isolatable;
+import org.gradle.internal.snapshot.ValueSnapshot;
+import org.gradle.internal.snapshot.ValueSnapshottingException;
+
+import javax.annotation.Nullable;
+
+import static org.gradle.internal.snapshot.impl.AbstractValueProcessor.javaSerialized;
+import static org.gradle.internal.snapshot.impl.JavaSerializedValueSnapshot.javaDeserialized;
+
+@NonNullApi
+public class IsolatedArrayOfPrimitive implements Isolatable<Object> {
+    private final PrimitiveType primitiveType;
+    private final Object array;
+
+    public IsolatedArrayOfPrimitive(Object array) {
+        this(PrimitiveType.of(array.getClass()), array);
+    }
+
+    private IsolatedArrayOfPrimitive(PrimitiveType primitiveType, Object array) {
+        this.primitiveType = primitiveType;
+        this.array = array;
+    }
+
+    public static IsolatedArrayOfPrimitive fromByteArray(byte primitiveTypeCode, byte[] bytes) {
+        PrimitiveType primitiveType = PrimitiveType.fromOrdinal(primitiveTypeCode);
+        return new IsolatedArrayOfPrimitive(
+            primitiveType,
+            fromByteArray(bytes, primitiveType)
+        );
+    }
+
+    @Override
+    public void appendToHasher(Hasher hasher) {
+        throw unsupportedOperation();
+    }
+
+    @Override
+    public ValueSnapshot asSnapshot() {
+        throw unsupportedOperation();
+    }
+
+    private UnsupportedOperationException unsupportedOperation() {
+        return new UnsupportedOperationException("This operation is not supported on primitive arrays of type: " + primitiveType.arrayType);
+    }
+
+    @Override
+    public Object isolate() {
+        return primitiveType.clone(array);
+    }
+
+    @Nullable
+    @Override
+    public <S> S coerce(Class<S> type) {
+        return null;
+    }
+
+    public byte getPrimitiveTypeCode() {
+        return (byte) primitiveType.ordinal();
+    }
+
+    public byte[] toByteArray() {
+        switch (primitiveType) {
+            case B:
+                return (byte[]) array;
+            default:
+                return javaSerialized(array);
+        }
+    }
+
+    private static Object fromByteArray(byte[] bytes, PrimitiveType primitiveType) {
+        switch (primitiveType) {
+            case B:
+                return bytes;
+            default:
+                return javaDeserialized(primitiveType.arrayType, bytes);
+        }
+    }
+
+    private enum PrimitiveType {
```

Comment:
❓ Do we miss boolean here or I missed it?

Reply:
Indeed!

Reply:
FTR: https://github.com/gradle/gradle/pull/27626